File: Data Structures/Graph/Graph.java
import java.util.Scanner;
import java.util.Arrays;
import java.util.InputMismatchException;


public class Graph {


    int adjacencyMatrix[][];
    int size;


    Graph(int v)
    {
        adjacencyMatrix= new int[v][v];
        this.size= v;
    }

    // Add edge between two vertices
    void addEdge(int to, int from)
    {  
        if (to>=size|| from>=size || to<0 || from<0)
            System.out.println("Invalid vertices");
        else{
        adjacencyMatrix[to][from]= 1;
        adjacencyMatrix[from][to]= 1;}
    }


    // Remove edge between two vertices
    void removeEdge(int to, int from) {

        if (to>=size || from>=size|| to<0 || from<0)
            System.out.println("Invalid vertices");

        else if (adjacencyMatrix[to][from]==0 || adjacencyMatrix[from][to]==0)
            System.out.println("Edge does not exist");
        else{
        

        adjacencyMatrix[to][from] = 0;
        adjacencyMatrix[from][to] = 0;}
    }

    
    // Display adjacency matrix
    void display(){

        for(int i=0;i<size;i++) {

            for (int j=0;j<size;j++)
            {
                System.out.print(adjacencyMatrix[i][j]+ " ");
                

            }
            System.out.println();

        }
    }

    // Perform DFS
    void dfs(int start, boolean visited[])
    {

        System.out.print(start + " ");
        visited[start]=true;

        for (int i=0;i<adjacencyMatrix[start].length;i++)
        {

            if (adjacencyMatrix[start][i]==1 && !visited[i]){
                dfs(i,visited);
            }


        }

    }

    // Perform BFS
    void bfs(int start)
    {

        // List<Integer> q= new ArrayList<>();
        Queue q= new Queue();

        boolean visited[]= new boolean [size];
        Arrays.fill(visited,false );

        // q.add(start);
        q.enqueue(start);

        visited[start]=true;

        int vis;

        while( !q.isEmpty())
        {
            vis= q.peek();

            System.out.print(vis +" ");

            q.dequeue();

            for(int i=0;i<size;i++){

                if (adjacencyMatrix[vis][i]==1 && !visited[i]){

                    q.enqueue(i);
                    visited[i]= true;

                }

            }



        }



    }


    public static void main(String[] args) {


        try{

        Scanner sc= new Scanner(System.in);

        System.out.println("Enter the number of vertices: ");
        int v= sc.nextInt();
        if (v<0)
            throw new Exception("Number of vertices cannot be negative");

        Graph g= new Graph(v);

        while(true){

            System.out.println("\n\n1. Add Edge \n2. Remove Edge \n3. Display Adjacency Matrix \n4. Perform DFS \n5. Perform BFS \n6. Exit ");

            System.out.println("\nEnter your choice: ");
            int ch= sc.nextInt();

            switch (ch){

                case 1:
                    System.out.println("\nEnter the vertices to add edge between: ");
                    int v1= sc.nextInt();
                    int v2= sc.nextInt();

                    g.addEdge(v1,v2);
                    break;

                case 2:
                System.out.println("\nEnter the vertices to remove edge between: ");
                int vt1= sc.nextInt();
                int vt2= sc.nextInt();

                g.removeEdge(vt1,vt2);
                break;

                case 3:
                    System.out.println("\nAdjacency Matrix: ");
                    g.display();
                    break;

                case 4:
                    System.out.println("\nEnter the starting vertex for DFS: ");
                    int start= sc.nextInt();
                    boolean visited[]= new boolean[v];
                    g.dfs(start,visited);
                    break;

                case 5:
                    System.out.println("\nEnter the starting vertex for BFS: ");
                    int startb= sc.nextInt();
                    g.bfs(startb);
                    break;

                case 6:
                    System.exit(0);
            }
        }
    }
    catch(InputMismatchException e){
        System.out.println("Invalid input, Please Enter a integer input. " );
        System.out.println("Exception: " + e + "");
    }

    catch(Exception e){
        System.out.println("Exception: " + e + "");
    }
}}

File: Data Structures/Graph/Graph.cpp
#include <iostream>
using namespace std;
// Class Queue
class Queue
{
public:
    int front, rear, capacity;
    int queue[1000];

    // Constructor 
    Queue(int capacity)
    {
        front = rear = 0;
        this->capacity = capacity ;
    }
    // Method push 
    void Push(int data)
    {
        // queue is full condition
        if (capacity == rear)
        {
            cout << "Queue is full" << endl;
            return;
        }

        else
        {
            queue[rear] = data;
            rear++;
        }
        return;
    }
    // Method pop 
    void Pop()
    {
        // Queue is empty
        if (front == rear)
        {
            return;
        }
        else
        {
           
            for (int i = 0; i < rear - 1; i++)
            {
                queue[i] = queue[i + 1];
            }
            rear--;
        }
    }
    
    int Front()
    {
        // Queue is empty condition
        if (front == rear)
        {
            return -1;
        }
        return queue[front];
    }
    bool Empty()
    {
        return front == rear;
    }
};
// Class Graph
class Graph
{
public:
    int vertices; 
    // create an Adjacencey matrix
    int **adjacencyMatrix;


    int firstNode;
    Graph(int value)
    {
        vertices = value;
        adjacencyMatrix = new int *[value + 1];
      
        for (int i = 0; i <= value ; i++)
        {
            adjacencyMatrix[i] = new int[value];
        }
        firstNode = -1; 
    }

    // Method to add Edge
    void addEdge()
    {
        int node1, node2;
        cout << "Enter the Value of Vertex1" << endl;
        cin >> node1;
        cout << "Enter the Value of Vertex2" << endl;
        cin >> node2;

        adjacencyMatrix[node1][node2] = 1;
        adjacencyMatrix[node2][node1] = 1;

        if (firstNode == -1) 
        {
            firstNode = node1;
        }
    }

    // Graph Traversal BFS
    void bfs()
    {
        // mark visited
        int visited[vertices + 1] = {0};


        Queue q(100);
        q.Push(firstNode);
        // mark visited
        visited[firstNode] = 1;

        while (!q.Empty())
        {
            int vertex = q.Front();
            cout << vertex << " ";
            q.Pop();
            // i = index
            for (int i = 0; i <= vertices; i++)
            {
                if (adjacencyMatrix[vertex][i] == 1 && !visited[i])
                {

                   
                    q.Push(i);

                   
                    visited[i] = 1;
                }
            }
        }
        cout << endl;
    }

    void DfsTraversal(int source, int visited[])
    {
        visited[source] = 1;
        cout << source << " ";

        for (int child = 0; child <= vertices; child++)
        {
            if (adjacencyMatrix[source][child] == 1 && !visited[child])
            {
                DfsTraversal(child, visited);
            }
        }
        return;
    }

    // Graph Traversal DFS
    void dfs()
    {
        // made a visited array to mark nodes visited
        int visited[vertices + 1] = {0};
        DfsTraversal(firstNode, visited);
        cout << endl;
        return;
    }

    // function to delete a Edge from Graph
    void deleteEdge()
    {
        int node1, node2;
        cout << "Enter Node1 and Node2 whose edge between them is to be deleted" << endl;
        cin >> node1 >> node2;

        if (node1 > vertices || node2 > vertices)
        {
            cout << "Invalid Node Values" << endl;
            return;
        }
      
        adjacencyMatrix[node1][node2] = 0;
        adjacencyMatrix[node2][node1] = 0;
        return;
    }

    // Function to Display Graph
    void display()
    {
        cout << "Your List is " << endl;
        for (int row = 0; row <= vertices; row++)
        {
            cout << row << "->";
            for (int column = 0; column <= vertices; column++)
            {
                if (adjacencyMatrix[row][column] == 1)
                {
                    cout << column << " ";
                }
            }
            cout << endl;
        }
    }
};

int main()
{
    int vertices;
    cout << "Enter Number of vertices" << endl;
    cin >> vertices;

   
    Graph g(vertices);
    int choice = 0;

    while (choice != 5)
    {
        cout << "\n\nLinked List Menu\n";
        cout << "==============================\n";
        cout << "1. Add an Edge" << endl;
        cout << "2. Traverse the Graph" << endl;
        cout << "3. Delete the Edge" << endl;
        cout << "4. Display" << endl;
        cout << "5. Exit" << endl;
        cout << "Enter Your Choice" << endl;
        cin >> choice;

        // Menu
        switch (choice)
        {
        case 1:
        {
            g.addEdge();
            break;
        }
        case 2:
        {
            cout << "Traversal Techniques" << endl;
            cout << "1. DFS" << endl;
            cout << "2. BFS" << endl;
            int choice1;
            cin >> choice1;
            if (choice1 == 1)
            {
                g.dfs();
            }
            else if (choice1 == 2)
            {
                g.bfs();
            }
            else
            {
                cout << "Invalid Choice" << endl;
            }
            break;
        }
        case 3:
        {
            g.deleteEdge();
            break;
        }
        case 4:
        {
            g.display();
            break;
        }
        case 5:
        {
            cout << "Exit" << endl;
            break;
        }
        default:
             cout<<"enter a valid number";
        }
    }
   
}